src/mlpack/core/data/save_impl.hpp

```diff
@@ -38,220 +39,177 @@ bool Save(const std::string& filename,
+      << "'." << std::endl;
+  if constexpr (IsArma<MatType>::value || IsSparseMat<MatType>::value)
   {
-    saveType = FileType::ArmaBinary;
-    stringType = "Armadillo binary formatted data";
+    TextOptions txtOpts(std::move(opts));
+    if constexpr (IsSparseMat<MatType>::value)
+    {
+      success = SaveSparse(matrix, txtOpts, stream);
+    }
+    else if constexpr (IsCol<MatType>::value)
+    {
+      opts.NoTranspose() = true;
+      success = SaveDense(matrix, txtOpts, stream);
+    }
+    else if constexpr (IsRow<MatType>::value)
+    {
+      opts.NoTranspose() = false;
+      success = SaveDense(matrix, txtOpts, stream);
+    }
+    else if constexpr (IsDense<MatType>::value)
+    {
+      success = SaveDense(matrix, txtOpts, stream);
+    }
+    opts = std::move(txtOpts);
   }
   else
   {
-    unknownType = true;
-    saveType = FileType::RawBinary; // Won't be used; prevent a warning.
-    stringType = "";
+    throw std::runtime_error("DataOptionType is unknown!."
+        "please use a known type or provide specific overloads");
   }
 
-  // Provide error if we don't know the type.
-  if (unknownType)
+  if (!success)
   {
     Timer::Stop("saving_data");
-    if (fatal)
-      Log::Fatal << "Unable to determine format to save to from filename '"
-          << filename << "'.  Save failed." << std::endl;
+    if (opts.Fatal())
+      Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
     else
-      Log::Warn << "Unable to determine format to save to from filename '"
-          << filename << "'.  Save failed." << std::endl;
-
+      Log::Warn << "Save to '" << filename << "' failed." << std::endl;
     return false;
   }
 
-  // Try to save the file.
-  Log::Info << "Saving " << stringType << " to '" << filename << "'."
-      << std::endl;
+  Timer::Stop("saving_data");
 
-  arma::SpMat<eT> tmp = matrix;
+  return success;
+}
 
+template<typename eT>
+bool SaveDense(const arma::Mat<eT>& matrix,
+               TextOptions& opts,
+               std::fstream& stream)
+{
+  bool success = false;
+  arma::Mat<eT> tmp;
   // Transpose the matrix.
-  if (transpose)
+  if (!opts.NoTranspose())
   {
     tmp = trans(matrix);
+    success = SaveMatrix(tmp, opts, stream);
   }
+  else
+    success = SaveMatrix(matrix, opts, stream);
 
-  const bool success = tmp.save(stream, ToArmaFileType(saveType));
-  if (!success)
-  {
-    Timer::Stop("saving_data");
-    if (fatal)
-      Log::Fatal << "Save to '" << filename << "' failed." << std::endl;
-    else
-      Log::Warn << "Save to '" << filename << "' failed." << std::endl;
+  return success;
+}
 
-    return false;
-  }
+// Save a Sparse Matrix
+template<typename eT>
+bool SaveSparse(const arma::SpMat<eT>& matrix,
+                TextOptions& opts,
+                std::fstream& stream)
+{
+  bool success = false;
+  arma::SpMat<eT> tmp;
 
-  Timer::Stop("saving_data");
+  // Transpose the matrix.
+  if (!opts.NoTranspose())
+  {
+    tmp = trans(matrix);
```

Comment:
```suggestion
    arma::SpMat<eT> tmp = trans(matrix);
```

Then there's no need to declare `tmp` outside this scope... really trivial cleanup.